### PR TITLE
Add documentation for shoot access

### PR DESCRIPTION
**Pull request by ary1992**
**How to categorize this PR?**

/area documentation
/kind enhancement

**What this PR does / why we need it**:
This PR adds documentation for accessing shoot with static token kubeconfig and dynamically created kubeconfig using shoot/adminKubeconfig subresource.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
Documentation for accessing the shoot cluster is added [here](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_access.md).
```

### CODE REVIEW

File: docs/usage/shoot_access.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# Shoot_access.md
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
# Accessing Shoot Clusters
```

Please also add this document to the index (`docs/README.md`).

File: docs/usage/shoot_access.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+There are several scenarios where end-user would like to access shoot cluster. There are several options available to user to access the shoot cluster.
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
After creation of a shoot clusters, end-users require a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.
```

File: docs/usage/shoot_access.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+## Static token kubeconfig :
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
## Static Token Kubeconfig
```

File: docs/usage/shoot_access.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  - To access the shoot cluster with the static token kubeconfig, shoot should have `.spec.kubernetes.enableStaticTokenKubeconfig` option set to `true`. enableStaticTokenKubeconfig will allow the creation of kubeconfig secret in the project namespace. End users can fetch the secret and use the kubeconfig inside it. Static token belongs to `system:masters` group and grants `cluster-admin` privileges to cluster.
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
This `kubeconfig` contains a [static token](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#static-token-file) and provides `cluster-admin` privileges.
It is created by default and persisted in the `<shoot-name>.kubeconfig` secret in the project namespace in the garden cluster.
```

File: docs/usage/shoot_access.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    spec:
+      kubernetes:
+        version: 1.23.1
+        enableStaticTokenKubeconfig: true
+      ...
+    ```
+  This is not the recommended method to access the shoot cluster as the static token kubeconfig has some security flaws associated with it.
+
+   - Static token in the kubeconfig doesn't have any expiration date. To revoke the static token, the user needs to rotate the kuebcconfig credentials.
+   - Static token doesn't have user identity associated with it. User in that token will always be system:cluster-admin irrespective of the person accessing the cluster. Hence it is impossible to audit the events in cluster.
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
...
spec:
  kubernetes:
    enableStaticTokenKubeconfig: true
...
\```

It is **not** the recommended method to access the shoot cluster as the static token `kubeconfig` has some security flaws associated with it:

- The static token in the `kubeconfig` doesn't have any expiration date. To revoke the static token, the user needs to rotate the kuebcconfig credentials (see https://github.com/gardener/gardener/blob/master/docs/usage/shoot_operations.md#rotate-kubeconfig-credentials).
- The static token doesn't have any user identity associated with it. The user in that token will always be `system:cluster-admin` irrespective of the person accessing the cluster. Hence, it is impossible to audit the events in cluster.
```

File: docs/usage/shoot_access.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    metadata:
+      name: local
+      namespace: garden-local
+      ...
+    spec:
+      kubernetes:
+        version: 1.23.1
+        enableStaticTokenKubeconfig: true
+      ...
+  ```
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
```

File: docs/usage/shoot_access.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+## Dynamic kubeconfig
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
## `shoots/adminkubeconfig` subresource
```

File: docs/usage/shoot_access.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+  Shoot subresource called `adminKubeconfig` allows user to dynamically generate short lived `kubeconfig` that can be used to access shoot cluster with `cluster-admin` priviledge. `shoots/adminkubeconfig` resource can only accept `CREATE` calls and accept   `AdminKubeconfigRequest` .Kubeconfig is generated only when `AdminKubeconfigRequest` sent to the subresource.
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
The `shoots/adminkubeconfig` subresource allows users to dynamically generate short-lived `kubeconfig`s that can be used to access shoot cluster with `cluster-admin` privileges. The credentials associated with this `kubeconfig` are client certificates which have a very short validity and must be renewed before they expire.
```

File: docs/usage/shoot_access.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  {
+      "apiVersion": "authentication.gardener.cloud/v1alpha1",
+      "kind": "AdminKubeconfigRequest",
+      "spec": {
+          "expirationSeconds": 1000
+      }
+  }
+  ```
+
+
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
> The [`gardenctl-v2`](https://github.com/gardener/gardenctl-v2/) tool makes it easy to target shoot clusters and automatically renews such `kubeconfig`s when required.
```

File: docs/usage/shoot_access.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        version: 1.23.1
+        enableStaticTokenKubeconfig: true
+      ...
+  ```
+
+## Dynamic kubeconfig
+
+  Shoot subresource called `adminKubeconfig` allows user to dynamically generate short lived `kubeconfig` that can be used to access shoot cluster with `cluster-admin` priviledge. `shoots/adminkubeconfig` resource can only accept `CREATE` calls and accept   `AdminKubeconfigRequest` .Kubeconfig is generated only when `AdminKubeconfigRequest` sent to the subresource.
+
+  User identity of kubeconfig will be the user authenticating to gardener API server. Following command can be used to request a kubeconfig using the `AdminKubeconfigRequest`.
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
The username associated with such `kubeconfig` will be the same which is used for authenticating to the Gardener API. Apart from this advantage, the created `kubeconfig`s will not be persisted anywhere.

The following command can be used to request a `kubeconfig` using the `AdminKubeconfigRequest` resource.
```

File: docs/usage/shoot_access.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  ```json
+  {
+      "apiVersion": "authentication.gardener.cloud/v1alpha1",
+      "kind": "AdminKubeconfigRequest",
+      "spec": {
+          "expirationSeconds": 1000
+      }
+  }
+  ```
+
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
```bash
export NAMESPACE=my-namespace
export SHOOT_NAME=my-shoot
kubectl create \
    -f example/shoot-kubeconfig/kubeconfig-request.json \
    --raw /apis/core.gardener.cloud/v1beta1/namespaces/${NAMESPACE}/shoots/${SHOOT_NAME}/adminkubeconfig | jq -r ".status.kubeconfig" | base64 -d
\```

Here, the `kubeconfig-request.json` has the following content:

```json
{
    "apiVersion": "authentication.gardener.cloud/v1alpha1",
    "kind": "AdminKubeconfigRequest",
    "spec": {
        "expirationSeconds": 1000
    }
}
\```
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
- We can now drop https://github.com/gardener/gardener/tree/master/example/shoot-kubeconfig, I think
- Can you add a link to this document here (https://github.com/gardener/gardener/blob/master/docs/usage/shoot_operations.md#rotate-kubeconfig-credentials)
- Can you add a third section talking about OIDC as option? https://github.com/gardener/gardener/blob/5023d12c26ddd05fb7f4ddbf40e3425f8176f462/example/90-shoot.yaml#L140

File: docs/usage/shoot_access.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
+# Accessing Shoot Clusters
+
+After creation of a shoot clusters, end-users requires a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.
+
+## Static Token Kubeconfig :
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
## Static Token Kubeconfig
```

File: docs/usage/shoot_access.md [hunk 13]
```diff
@@ -0,0 +1,58 @@
+# Accessing Shoot Clusters
+
+After creation of a shoot clusters, end-users requires a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
After creation of a shoot clusters, end-users require a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.
```

File: docs/usage/shoot_access.md [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+This `kubeconfig` contains a [static token](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#static-token-file) and provides `cluster-admin` privileges.
+It is created by default and persisted in the `<shoot-name>.kubeconfig` secret in the project namespace in the garden cluster.
+
+```yaml
+apiVersion: core.gardener.cloud/v1beta1
+kind: Shoot
+...
+spec:
+kubernetes:
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
  kubernetes:
```

File: docs/usage/shoot_access.md [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    enableStaticTokenKubeconfig: true
+...
+```
+It is **not** the recommended method to access the shoot cluster as the static token `kubeconfig` has some security flaws associated with it:
```
Comment (rfranzke) [inline docs/usage/shoot_access.md:20]:
```suggestion

It is **not** the recommended method for accessing the shoot cluster as the static token `kubeconfig` has some security flaws associated with it:
```

File: docs/usage/shoot_access.md [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+- The static token in the `kubeconfig` doesn't have any expiration date. To revoke the static token, the user needs to rotate the kuebcconfig credentials (see https://github.com/gardener/gardener/blob/master/docs/usage/shoot_operations.md#rotate-kubeconfig-credentials).
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
- The static token in the `kubeconfig` doesn't have any expiration date. To revoke the static token, the user needs to rotate the kubeconfig credentials (see https://github.com/gardener/gardener/blob/master/docs/usage/shoot_operations.md#rotate-kubeconfig-credentials).
```

File: docs/usage/shoot_access.md [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+- The static token doesn't have any user identity associated with it. The user in that token will always be `system:cluster-admin` irrespective of the person accessing the cluster. Hence, it is impossible to audit the events in cluster.
+
+## `shoots/adminkubeconfig` subresource
+
+The [`shoots/adminkubeconfig`](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md) subresource allows users to dynamically generate short-lived `kubeconfig` that can be used to access shoot cluster with `cluster-admin` privileges. The credentials associated with this `kubeconfig` are client certificates which have a very short validity and must be renewed before they expire.
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
The [`shoots/adminkubeconfig`](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md) subresource allows users to dynamically generate temporary `kubeconfig`s that can be used to access shoot cluster with `cluster-admin` privileges. The credentials associated with this `kubeconfig` are client certificates which have a very short validity and must be renewed before they expire (by calling the subresource  endpoint again).
```

File: docs/usage/shoot_access.md [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+The username associated with such `kubeconfig` will be the same which is used for authenticating to the Gardener API. Apart from this advantage, the created `kubeconfig` will not be persisted anywhere.
+
```
Comment (rfranzke) [inline docs/usage/shoot_access.md:29]:
```suggestion

In order to request such a `kubeconfig`, you can run the following commands:

```

File: docs/usage/shoot_access.md [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+```json
+{
+    "apiVersion": "authentication.gardener.cloud/v1alpha1",
+    "kind": "AdminKubeconfigRequest",
+    "spec": {
+        "expirationSeconds": 1000
+    }
+}
+```
+> The [`gardenctl-v2`](https://github.com/gardener/gardenctl-v2/) tool makes it easy to target shoot clusters and automatically renews such `kubeconfig` when required.
```
Comment (rfranzke) [inline docs/usage/shoot_access.md:52]:
```suggestion

> The [`gardenctl-v2`](https://github.com/gardener/gardenctl-v2/) tool makes it easy to target shoot clusters and automatically renews such `kubeconfig` when required.
```

File: docs/usage/shoot_access.md [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+## `oidcConfig`
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
## OpenID Connect
```

File: docs/usage/shoot_access.md [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+Gardener provides an admission controller (ClusterOpenIDConnectPreset and OpenIDConnectPreset) which, when enabled, applies ClusterOpenIDConnectPresets and OpenIDConnectPresets respectivly to incoming `Shoot` creation requests having label `oidc : enable` set. It mutates the shoot's `.spec.kubernetes.kubeAPIServer.oidcConfig` field with the oidc configuration which will be used for  `kubeconfig` generation.
+
+It is the end-user responsibility to incorporate the OpenID Connect configurations in kubeconfig to access the cluster. One way is to use the kubectl plugin called [`kubectl oidc-login`](https://github.com/int128/kubelogin)  for OIDC authentication.
+
+For further information on (Cluster)OpenIDConnectPreset, refer to [doc](https://github.com/gardener/gardener/blob/master/docs/usage/openidconnect-presets.md).
```
Comment (rfranzke) [inline docs/usage/shoot_access.md]:
```suggestion
The `kube-apiserver` of shoot clusters can be provided with [OpenID Connect configuration](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#openid-connect-tokens) via the `ShootSpec`:

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
...
spec:
  kubernetes:
    oidcConfig:
      ...
\```

It is the end-user's responsibility to incorporate the OpenID Connect configurations in `kubeconfig` for accessing the cluster (i.e., Gardener will not automatically generate `kubeconfig` based on these OIDC settings).
The recommended way is using the `kubectl` plugin called [`kubectl oidc-login`](https://github.com/int128/kubelogin) for OIDC authentication.

If you want to use the same OIDC configuration for all your shoots by default then you can use the `ClusterOpenIDConnectPreset` and `OpenIDConnectPreset` API resources. They allowing defaulting the `.spec.kubernetes.kubeAPIServer.oidcConfig` fields for newly created `Shoot`s such that you don't have to repeat yourself every time (similar to `PodPreset` resources in Kubernetes).
`ClusterOpenIDConnectPreset` specified OIDC configuration applies to `Projects` and `Shoots` cluster-wide (hence, only available to Gardener operators) while `OpenIDConnectPreset` is `Project`-scoped.
Shoots have to "opt-in" for such defaulting by using the `oidc=enable` label.

For further information on `(Cluster)OpenIDConnectPreset`, refer to [this document](https://github.com/gardener/gardener/blob/master/docs/usage/openidconnect-presets.md).
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
Apart from the inline suggestions, can you add a link to your document to this section at the end: https://github.com/gardener/gardener/blob/master/docs/usage/shoot_operations.md#rotate-kubeconfig-credentials

**Comment by rfranzke** [comment]
Can you also please provide a proper release note for users so that they get aware of this document?

File: docs/usage/shoot_access.md [hunk 22]
```diff
@@ -0,0 +1,75 @@
+# Accessing Shoot Clusters
+
+After creation of a shoot clusters, end-users require a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.
```
Comment (acumino) [inline docs/usage/shoot_access.md]:
```suggestion
After creation of a shoot cluster, end-users require a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.
```

File: docs/usage/shoot_access.md [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+It is **not** the recommended method to access the shoot cluster as the static token `kubeconfig` has some security flaws associated with it:
+- The static token in the `kubeconfig` doesn't have any expiration date. To revoke the static token, the user needs to rotate the kubeconfig credentials (see [rotate-kubeconfig-credentials](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_operations.md#rotate-kubeconfig-credentials)).
+- The static token doesn't have any user identity associated with it. The user in that token will always be `system:cluster-admin` irrespective of the person accessing the cluster. Hence, it is impossible to audit the events in cluster.
+
+## `shoots/adminkubeconfig` subresource
+
+The [`shoots/adminkubeconfig`](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md) subresource allows users to dynamically generate temporary `kubeconfig`s that can be used to access shoot cluster with `cluster-admin` privileges. The credentials associated with this `kubeconfig` are client certificates which have a very short validity and must be renewed before they expire (by calling the subresource  endpoint again).
```
Comment (acumino) [inline docs/usage/shoot_access.md]:
```suggestion
The [`shoots/adminkubeconfig`](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md) subresource allows users to dynamically generate temporary `kubeconfig`s that can be used to access shoot cluster with `cluster-admin` privileges. The credentials associated with this `kubeconfig` are client certificates which have a very short validity and must be renewed before they expire (by calling the subresource endpoint again).
```

File: docs/usage/shoot_access.md [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+It is the end-user's responsibility to incorporate the OpenID Connect configurations in `kubeconfig` for accessing the cluster (i.e., Gardener will not automatically generate `kubeconfig` based on these OIDC settings).
+The recommended way is using the `kubectl` plugin called [`kubectl oidc-login`](https://github.com/int128/kubelogin) for OIDC authentication.
+
+If you want to use the same OIDC configuration for all your shoots by default then you can use the `ClusterOpenIDConnectPreset` and `OpenIDConnectPreset` API resources. They allow defaulting the `.spec.kubernetes.kubeAPIServer.oidcConfig` fields for newly created `Shoot`s such that you don't have to repeat yourself every time (similar to `PodPreset` resources in Kubernetes).
+`ClusterOpenIDConnectPreset` specified OIDC configuration applies to `Projects` and `Shoots` cluster-wide (hence, only available to Gardener operators) while `OpenIDConnectPreset` is `Project`-scoped.
+Shoots have to "opt-in" for such defaulting by using the `oidc=enable` label.
+
+For further information on `(Cluster)OpenIDConnectPreset`, refer to [this document](https://github.com/gardener/gardener/blob/master/docs/usage/openidconnect-presets.md).
```
Comment (acumino) [inline docs/usage/shoot_access.md]: I guess the relative link is the preferred way. Similar changes in other places.
```suggestion
For further information on `(Cluster)OpenIDConnectPreset`, refer to [this document](openidconnect-presets.md).
```

**Review by acumino — COMMENTED** [review]
Minor Nits

File: docs/usage/shoot_access.md (hunk 24 above)
Comment (acumino) [reply docs/usage/shoot_access.md]: /cc @rfranzke

**Review by rfranzke — COMMENTED** [review]
Thanks!

/approve

**Pull request merged**